src/control/lib/daos/client/api.go

```diff
@@ -0,0 +1,289 @@
+package client
+
+import (
+	"context"
+	"io"
+	"sync"
+	"unsafe"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/lib/daos"
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+/*
+#cgo LDFLAGS: -ldaos -ldfs -lduns
+
+#include <daos.h>
+*/
+import "C"
+
+const (
+	apiCtxKey ctxKey = "apiCtxKey"
+)
+
+var (
+	log logging.DebugLogger = logging.NewDebugLogger(io.Discard)
+)
+
+type (
+	// APIContext is an opaque type used to convey DAOS context across API methods.
+	APIContext struct {
+		sync.RWMutex
+		cancelCtx context.CancelFunc
+		poolConns map[string]*ConnHandle
+		contConns map[string]map[string]*ConnHandle
```

Comment:
I agree with getting rid of this stuff. I think letting the client deal with the handles directly is probably the best way to go.

I do think the context object is a nice way to handle daos_init and daos_fini.